File: experiments/java/sort/files/ListImplCCC.java
class ListImpl<T> implements IList<T>
{
	private ListNode first;	
	private long Size;
	public ListImpl(T val)
	{
		Size=1;
		first = new ListNode(val);
	}
	
	public void Add(T val)
	{
		ListNode newNode = new ListNode(val);
		newNode.prev = first.prev;
		newNode.next = first;
		first.prev = newNode;
		newNode.prev.next = newNode;
		Size=Size+1;
	}
	
	public IIterator<T> GetIterator()
	{
		IIterator<T> iter = this.MakeIterator(first);
		return iter;
	}
	public long GetSize()
	{
		return Size;
	}

	private class ListNode
	{
		public ListNode(T v)
		{
			value=v;
			next=this;
			prev=this;
		}
		public T value;
		public ListNode next;
		public ListNode prev;
	}

	public IIterator<T> MakeIterator(ListNode n)
	{
		ListImpl<T> self = this;
		return new IteratorImpl(self, n);
	}

	private class IteratorImpl implements IIterator<T>
	{
		ListNode currentNode;
		ListImpl<T> parent;
		public IteratorImpl(ListImpl<T> l, ListNode node)
		{
			currentNode = node;
			parent = l;
		}
	
		public boolean MovePrev()
		{
			if(currentNode==parent.first)
			{
				return false;
			}
			currentNode = currentNode.prev;
			return true;
		}
		public boolean MoveNext()
		{
			if(currentNode.next==parent.first)
			{
				return false;
			}
			currentNode=currentNode.next;
			return true;
		}
		public T Current()
		{
			return currentNode.value;
		}
		public void SetValue(T x)
		{
			currentNode.value = x;
		}
		public IIterator<T> Clone()
		{
			return parent.MakeIterator(currentNode);
		}
	}
}

File: experiments/higgs/sort/files/listII/listimplCCC.ts
    class ListNode
    {
        value : number;
        next : ListNode;
        prev : ListNode;

        constructor(v : number)
        {
            this.value = v;
            this.next = this;
            this.prev = this;
        }
    }

    class List implements IList
    {
	    first : ListNode;
        private size : number;

        constructor(val : number)
        {
            this.size = 1;
            this.first = new ListNode(val);
        }
	
        add(val : number) : void
        {
            var newNode : ListNode = new ListNode(val);
            newNode.prev = this.first.prev;
            newNode.next = this.first;
            this.first.prev = newNode;
            newNode.prev.next = newNode;
            this.size = this.size + 1;
        }
	
        getIterator() : IIterator
        {
            var iter : IIterator = this.makeIterator(this.first);
            return iter;
        }
        
        getSize() : number
        {
            return this.size;
        }

        makeIterator(n : ListNode) : IIterator
        {
            var self : List = this;
            return new IteratorImpl(self, n);
        }
    }

	class IteratorImpl implements IIterator
	{
		private currentNode : ListNode;
		private parent : List;

		constructor(l : List, node : ListNode)
		{
			this.currentNode = node;
			this.parent = l;
		}
	
		movePrev() : boolean
		{
			if (this.currentNode === this.parent.first)
			{
				return false;
			}
			this.currentNode = this.currentNode.prev;
			return true;
		}

		moveNext() : boolean
		{
			if (this.currentNode.next === this.parent.first)
			{
				return false;
			}
			this.currentNode = this.currentNode.next;
			return true;
		}

		current() : number
		{
			return this.currentNode.value;
		}

		setValue(x : number) : void
		{
			this.currentNode.value = x;
		}

		clone() : IIterator
		{
			return this.parent.makeIterator(this.currentNode);
		}
	}

